src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

```diff
@@ -0,0 +1,514 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace Microsoft.Collections.Extensions
+{
+    internal enum InsertionBehavior
+    {
+        None = 0,
+        OverwriteExisting = 1,
+        ThrowOnExisting = 2
+    }
+
+    [DebuggerTypeProxy(typeof(IDictionaryDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public partial class OrderedDictionary<TKey, TValue> : IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue>, IList<KeyValuePair<TKey, TValue>>, IReadOnlyList<KeyValuePair<TKey, TValue>>
+    {
+        private struct Entry
+        {
+            public int HashCode; // lower 31 bits of the hash code
+            public TKey Key;
+            public TValue Value;
+            public int Next; // the index of the next item in the same bucket
+        }
+
+        private int[] _buckets = Array.Empty<int>();
```

Comment:
Any particular reason you chose to initialize by default to capacity 0, compared with `Dictionary<TKey, TValue>` which looks like it will never have arrays smaller than 3? 

Although I wonder whether Dictionary should do the same. It makes empty ones cheaper.

Reply:
Correction, I see Dictionary starts at 0 with null arrays, you just exchange a test for null array against a test for empty array. Maybe it makes no difference.